File: src/clase-04/practica/python/singleton_starter.py
"""
Actividad 1 — Singleton (Python)
Objetivo: Implementar un Logger como Singleton.

Requisitos:
- Debe garantizar UNA instancia global.
- Método set_level(...) con niveles: INFO, DEBUG, WARN, ERROR.
- Método log(msg) que imprima: [NIVEL] mensaje
- Demostración: crear dos referencias y comprobar que apuntan a la misma instancia.
"""

class Logger:
    _instance = None  # TODO: usar para guardar la única instancia

    def __new__(cls):
        # TODO: implementar patrón Singleton aquí
        return super().__new__(cls)

    def __init__(self):
        # TODO: inicializar nivel por defecto (INFO) solo la primera vez
        pass

    def set_level(self, level: str) -> None:
        # TODO: validar y setear el nivel
        pass

    def log(self, message: str) -> None:
        # TODO: imprimir con el formato [NIVEL] mensaje
        pass


if __name__ == "__main__":
    # TODO: crear 2 referencias al Logger y demostrar que son la misma instancia
    # TODO: cambiar el nivel en una referencia y loguear desde la otra
    pass

File: src/clase-04/practica/typescript/singleton_starter.ts
// Actividad 1 — Singleton (TypeScript)
// Objetivo: Implementar un Logger como Singleton.
//
// Requisitos:
// - ÚNICA instancia global (propiedad estática + constructor privado)
// - setLevel(level) con niveles: INFO | DEBUG | WARN | ERROR
// - log(message): imprime `[NIVEL] mensaje`
// - Demostración: dos referencias apuntan a la misma instancia

type Level = "INFO" | "DEBUG" | "WARN" | "ERROR";

class Logger {
  // TODO: propiedad estática para guardar la instancia
  // private static instance: Logger;

  // TODO: estado interno (nivel actual)
  // private level: Level = "INFO";

  // TODO: constructor privado
  // private constructor() {}

  static getInstance(): Logger {
    // TODO: crear y/o devolver la única instancia
    throw new Error("TODO: implementar getInstance()");
  }

  setLevel(level: Level): void {
    // TODO: validar y setear el nivel
  }

  log(message: string): void {
    // TODO: imprimir formato [NIVEL] mensaje
  }
}

// Uso de ejemplo (descomentar al completar)
// const l1 = Logger.getInstance();
// const l2 = Logger.getInstance();
// console.log("¿Misma instancia?", l1 === l2);
// l1.setLevel("DEBUG");
// l2.log("Sistema iniciado");
